### PR TITLE
Implement Inspect protocol for BSON types

### PR DESCRIPTION
**Pull request by michalmuskala**
It probably could be better for certain types, but I think it's better then raw structs.

### CODE REVIEW

File: lib/bson/types.ex [hunk 1]
```diff
@@ -12,20 +22,54 @@ defmodule BSON.ObjectId do
...
   defstruct [:code, :scope]
+
+  defimpl Inspect do
+    def inspect(%BSON.JavaScript{code: code, scope: nil}, _opts) do
+      "#BSON.JavaScript<#{code}>"
+    end
+
+    def inspect(%BSON.JavaScript{code: code, scope: scope}, _opts) do
+      "#BSON.JavaScript<#{code}, #{inspect(scope)}>"
+    end
```
Comment (ericmj) [inline lib/bson/types.ex:92]: Javascript code needs to be inspected so we get quotes around and escaping.

File: lib/bson/types.ex [hunk 2]
```diff
@@ -1,5 +1,15 @@
 defmodule BSON.Binary do
   defstruct [binary: nil, subtype: :generic]
+
+  defimpl Inspect do
+    def inspect(%BSON.Binary{binary: value, subtype: :generic}, _opts) do
+      "#BSON.Binary<#{Base.encode16(value, case: :lower)}>"
+    end
+
+    def inspect(%BSON.Binary{binary: value, subtype: subtype}, _opts) do
+      "#BSON.Binary(#{subtype})<#{Base.encode16(value, case: :lower)}>"
```
Comment (ericmj) [inline lib/bson/types.ex]: Can we print this as: `#BSON.Binary<#{Base.encode16(value, case: :lower)}, #{subtype}>`

File: lib/bson/types.ex [hunk 3]
```diff
@@ -12,20 +22,54 @@ defmodule BSON.ObjectId do
...
+    end
+  end
 end
 
 defmodule BSON.Regex do
   defstruct [:pattern, :options]
+
+  defimpl Inspect do
+    def inspect(%BSON.Regex{pattern: pattern, options: options}, _opts) do
+      "#BSON.Regex/#{pattern}/#{options}"
```
Comment (ericmj) [inline lib/bson/types.ex]: Please print this as: `#BSON.Regex<#{inspect pattern}, #{inspect options}>`.

File: lib/bson/types.ex [hunk 4]
```diff
@@ -12,20 +22,54 @@ defmodule BSON.ObjectId do
...
+  end
 end
 
 defmodule BSON.DateTime do
   defstruct [:utc]
+
+  defimpl Inspect do
+    def inspect(%BSON.DateTime{utc: utc}, _opts) do
+      "#BSON.DateTime<#{utc}>"
+    end
```
Comment (ericmj) [inline lib/bson/types.ex:64]: Can we inspect this in the same format as ecto, iso8601 i think.

**Comment by ericmj** [comment]
Thank you, I just added a few comments.

**Comment by michalmuskala** [comment]
I changed everything as you asked and added two functions to the `BSON.DateTime` module based on conversions in Ecto and in my adapter, we could make them private, but I think it's fine to keep them there (especially that I need to convert `BSON.DateTime` struct to the datetime tuple in the adpter).

**Pull request merged**